aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp

```diff
@@ -134,87 +134,98 @@ void s_addmm_out_sparse_dense_worker(int64_t nnz, int64_t dim_i, int64_t dim_j,
+  TORCH_CHECK(((self_->dim() == 2) && (self_->sizes()[0] == mat1.sizes()[0]) && (self_->sizes()[1] == mat2.sizes()[1])),
+  "The input tensor must be a matrix with size ", mat1.sizes()[0], "x", mat2.sizes()[1], ", but got a ", self_->dim(),
+  "-D tensor with size ", self__sizes[0], "x", self__sizes[1]);
 
-  // ixj * jxk = ixk
-  int64_t dim_i = sparse.size(0);
-  int64_t dim_j = sparse.size(1);
-  int64_t dim_k = dense.size(1);
+  if (&result != &self) {
+    if (result.layout() == kStrided) {
+      at::native::resize_output(result, self__sizes);
+    } else {
+      at::native::resize_as_sparse_csr_(result, *self_);
+    }
+    result.copy_(*self_);
+  }
 
-  TORCH_CHECK(
-      dense.size(0) == dim_j,
-      "addmm: Expected dense matrix (op2) size(0)=",
-      dim_j,
-      ", got ",
-      dense.size(0));
+  IntArrayRef result_sizes = result.sizes();
+  if ((result_sizes[0] == 0) || (result_sizes[1] == 0)) {
+    return result;
+  }
 
-  resize_output(r, {dim_i, dim_k});
-  auto col_indices = sparse.col_indices();
-  auto crow_indices = sparse.crow_indices();
-  auto values = sparse.values();
-  int64_t nnz        = sparse._nnz();
-  if (nnz == 0) {
-    at::mul_out(r, t, at::scalar_tensor(beta, r.options()));
-    return r;
+  if (mat1._nnz() == 0 && mat2.layout() == kStrided) {
+    // According to docs, when beta==0 values in self should be ignored. nans and infs should not propagate
+    if (beta.toComplexDouble() == 0.) {
+      result.zero_();
+    } else {
+      result.mul_(beta);
+    }
+    return result;
   }
-  // Do not use MKL for Windows due to linking issues with sparse MKL routines.
-  if (at::hasMKL() && is_mkl_supported() && is_square_or_vec(dim_i, dim_j, dim_k)) {
-    AT_DISPATCH_FLOATING_TYPES(values.scalar_type(), "addmm_sparse_dense", [&] {
-        scalar_t cast_beta = beta.to<scalar_t>();
-        if (cast_beta == 0) {
-          r.zero_();
-        } else if (cast_beta == 1) {
-          if (!is_same_tensor(r, t)) {
-            r.copy_(t);
-          }
-        } else {
-          at::mul_out(r, t, scalar_to_tensor(beta));
-        }
-        // r = r + alpha * sparse * dense
-        _sparse_mm_mkl_(r, sparse, dense, t, alpha, Scalar(static_cast<scalar_t>(1.0)));
-    });
-  } else {
-    // r = beta * t + alpha * sparse * dense
-    AT_DISPATCH_FLOATING_TYPES(values.scalar_type(), "addmm_sparse_dense", [&] {
-        s_addmm_out_sparse_dense_worker<scalar_t>(nnz, dim_i, dim_j, dim_k, r, beta, t, alpha, crow_indices, col_indices, values, dense);
-    });
+
+  if (mat2.is_sparse_csr() && (mat1._nnz() == 0 || mat2._nnz() == 0)) {
+    if (beta.toComplexDouble() == 0.) {
+      result.values().zero_();
+    } else {
+      result.values().mul_(beta);
+    }
+    return result;
   }
-  return r;
+
+#if !AT_MKL_ENABLED()
+    if (mat2.is_sparse_csr() && result.is_sparse_csr()) {
+      TORCH_CHECK(false, "Calling addmm on sparse CPU tensors requires compiling PyTorch with MKL. Please use PyTorch built with MKL.");
```

Comment:
nit: I know this is using AT_MKL_ENABLED instead of AT_USE_MKL_SPARSE, but for this particular condition, shouldn't the error message be the same (i.e. it needs MKL *and* linux)?

Reply:
Right, use `AT_USE_MKL_SPARSE` should probably be used here instead of `AT_MKL_ENABLED`.